api/routes.go

```diff
@@ -3,31 +3,47 @@ package api
 import (
 	"context"
 	"net/http"
+	"time"
 
 	"github.com/julienschmidt/httprouter"
+	"github.com/sirupsen/logrus"
 )
 
 // buildHTTPRoutes registers all HTTP routes and their handlers.
 func (api *API) buildHTTPRoutes() {
-	api.staticRouter.GET("/user", validate(api.userHandlerGET))
+	api.staticRouter.GET("/user", api.validate(api.userHandler))
 	//api.staticRouter.PUT("/user", validate(api.userHandlerPUT))
-	//api.staticRouter.POST("/track/upload/:skylink", validate(api.trackUploadHandler))
-	//api.staticRouter.POST("/track/download/:skylink", validate(api.trackDownloadHandler))
+	api.staticRouter.GET("/user/uploads", api.validate(api.userUploadsHandler))
+	api.staticRouter.GET("/user/downloads", api.validate(api.userDownloadsHandler))
+	api.staticRouter.POST("/track/upload/:skylink", api.validate(api.trackUploadHandler))
+	api.staticRouter.POST("/track/download/:skylink", api.validate(api.trackDownloadHandler))
 }
 
 // validate ensures that the user making the request has logged in.
-func validate(h httprouter.Handle) httprouter.Handle {
+func (api *API) validate(h httprouter.Handle) httprouter.Handle {
 	return func(w http.ResponseWriter, req *http.Request, ps httprouter.Params) {
-		t, err := tokenFromRequest(req)
+		tokenStr, err := tokenFromRequest(req)
 		if err != nil {
-			WriteError(w, err, http.StatusBadRequest)
+			api.staticLogger.Traceln("error while fetching token from request", err)
+			WriteError(w, err, http.StatusUnauthorized)
 			return
 		}
-		token, err := ValidateToken(t)
+		token, err := ValidateToken(api.staticLogger, tokenStr)
 		if err != nil {
+			api.staticLogger.Traceln("error while validating token", err)
 			WriteError(w, err, http.StatusUnauthorized)
 			return
 		}
+		// If we don't have a valid cookie with reasonably long remaining TTL
+		// then set one.
+		c, err := req.Cookie(CookieName)
+		exp, _ := tokenExpiration(token)
+		if err != nil || !c.Expires.Equal(time.Unix(exp, 0)) {
```

Comment:
I don't understand this part of the condition:

```go
!c.Expires.Equal(time.Unix(exp, 0)
```

Isn't the chance that the two times are equal very low?

Reply:
You are right, I should be checking if the expiration time is *before* the current moment, no equal to it.
I'll fix it in a F/U.